File: src/main/java/com/mateusz/spring_boot_todos/controllers/UserRegistrationController.java
package com.mateusz.spring_boot_todos.controllers;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.factory.PasswordEncoderFactories;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.provisioning.JdbcUserDetailsManager;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.mateusz.spring_boot_todos.DTO.UserRegistrationRequest;
import com.mateusz.spring_boot_todos.DTO.UserRegistrationResponse;

import jakarta.validation.Valid;

import org.springframework.security.core.userdetails.User;

@RestController
@RequestMapping("/v1")
public class UserRegistrationController {

    private final JdbcUserDetailsManager userDetailsManager;

    public UserRegistrationController(JdbcUserDetailsManager userDetailsManager) {
        this.userDetailsManager = userDetailsManager;
    }

    @PostMapping(value = "/signup", consumes = { "application/json", "application/xml" }, produces = {
            "application/json",
            "application/xml" })
    public ResponseEntity<UserRegistrationResponse> registerUser(
            @Valid @RequestBody UserRegistrationRequest registrationRequest) {
        if (userDetailsManager.userExists(registrationRequest.username())) {
            return ResponseEntity.badRequest().body(new UserRegistrationResponse("Username is already taken!"));
        }

        PasswordEncoder encoder = PasswordEncoderFactories.createDelegatingPasswordEncoder();
        var user = User.withUsername(registrationRequest.username())
                .password(encoder.encode(registrationRequest.password()))
                .authorities("ROLE_USER")
                .build();

        userDetailsManager.createUser(user);

        return ResponseEntity.status(HttpStatus.CREATED)
                .body(new UserRegistrationResponse("User registered successfully!"));

    }
}

File: src/main/kotlin/com/mateusz/todos/kotlin_sboot_todos_backend/controllers/UserRegistrationController.kt
package com.mateusz.todos.kotlin_sboot_todos_backend.controllers

import com.mateusz.todos.kotlin_sboot_todos_backend.dto.UserRegistrationRequest
import com.mateusz.todos.kotlin_sboot_todos_backend.dto.UserRegistrationResponse
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.userdetails.User
import org.springframework.security.crypto.factory.PasswordEncoderFactories
import org.springframework.security.provisioning.JdbcUserDetailsManager
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/v1")
@Suppress("SpringJavaInjectionPointsAutowiringInspection")
class UserRegistrationController(private var userDetailsManager: JdbcUserDetailsManager) {
    @PostMapping(
        value = ["/signup"],
        consumes = ["application/json", "application/xml"],
        produces = ["application/json", "application/xml"]
    )
    fun registerUser(
        @RequestBody registrationRequest: @Valid UserRegistrationRequest
    ): ResponseEntity<UserRegistrationResponse> {
        if (userDetailsManager.userExists(registrationRequest.username)) {
            return ResponseEntity.badRequest().body(UserRegistrationResponse("Username is already taken!"))
        }

        val encoder = PasswordEncoderFactories.createDelegatingPasswordEncoder()
        val user = User.withUsername(registrationRequest.username)
            .password(encoder.encode(registrationRequest.password))
            .authorities("ROLE_USER")
            .build()

        userDetailsManager.createUser(user)

        return ResponseEntity.status(HttpStatus.CREATED)
            .body(UserRegistrationResponse("User registered successfully!"))
    }
}